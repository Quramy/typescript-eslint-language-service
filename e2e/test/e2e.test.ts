import path from "path";
import assert from "assert";
import ts from "typescript/lib/tsserverlibrary";
import { createServer, TSServer } from "../helper/server";

function findEventResponse(responses: any[], eventName: string) {
  return responses.find(response => response.event === eventName);
}

function findCommandResponse(responses: any[], commandName: string) {
  return responses.find(response => response.command === commandName);
}

function maskFileNameForDiagnostics(result: ts.server.protocol.DiagnosticEvent) {
  if (!result.body) return result;
  result.body.file = "<file>";
  return result;
}

function maskFileNameForCodeFixes(response: ts.server.protocol.CodeFixResponse) {
  if (!response.body) return response;
  response.body = response.body.map(b => {
    b.fixName = "<fileName>";
    b.changes = b.changes.map(c => {
      c.fileName = "<fileName>";
      return c;
    });
    return b;
  });
  return response;
}

describe("LanguageService plugin", () => {
  let server: TSServer;

  afterEach(async () => {
    if (server) await server.close();
  });
  describe("#getSemanticDiagnostics", () => {
    it("should not return ESLint error when the project does not use @typescript-eslint/parser", async () => {
      server = createServer({ projectPath: path.resolve(__dirname, "../projects/other-parser") });
      const { file, fileContent } = server.readFile("./main.ts");
      server.send({ command: "open", arguments: { file, fileContent, scriptKindName: "TS" } });
      await server.waitEvent("projectLoadingFinish");
      server.send({ command: "geterr", arguments: { files: [file], delay: 0 } });
      await server.waitEvent("semanticDiag");
      const found = findEventResponse(server.responses, "semanticDiag");
      if (!found) {
        throw new assert.AssertionError();
      }
      expect(maskFileNameForDiagnostics(found)).toMatchSnapshot();
    });

    it("should return ESLint error when the project uses @typescript-eslint/parser", async () => {
      server = createServer({ projectPath: path.resolve(__dirname, "../projects/simple") });
      const { file, fileContent } = server.readFile("./main.ts");
      server.send({ command: "open", arguments: { file, fileContent, scriptKindName: "TS" } });
      await server.waitEvent("projectLoadingFinish");
      server.send({ command: "geterr", arguments: { files: [file], delay: 0 } });
      await server.waitEvent("semanticDiag");
      const found = findEventResponse(server.responses, "semanticDiag");
      if (!found) {
        throw new assert.AssertionError();
      }
      expect(maskFileNameForDiagnostics(found)).toMatchSnapshot();
    });

    it("should return ESLint error when the project is configured with ESLint plugins", async () => {
      server = createServer({ projectPath: path.resolve(__dirname, "../projects/ts-eslint-plugin") });
      const { file, fileContent } = server.readFile("./main.ts");
      server.send({ command: "open", arguments: { file, fileContent, scriptKindName: "TS" } });
      await server.waitEvent("projectLoadingFinish");
      server.send({ command: "geterr", arguments: { files: [file], delay: 0 } });
      await server.waitEvent("semanticDiag");
      const found = findEventResponse(server.responses, "semanticDiag");
      if (!found) {
        throw new assert.AssertionError();
      }
      expect(maskFileNameForDiagnostics(found)).toMatchSnapshot();
    });

    it("should not reproduce issue #7", async () => {
      server = createServer({ projectPath: path.resolve(__dirname, "../projects/ts-eslint-plugin") });
      const { file, fileContent } = server.readFile("./reproduce_issue_7/main.ts");
      server.send({ command: "open", arguments: { file, fileContent, scriptKindName: "TS" } });
      await server.waitEvent("projectLoadingFinish");
      server.send({ command: "geterr", arguments: { files: [file], delay: 0 } });
      await server.waitEvent("semanticDiag");
      const found = findEventResponse(server.responses, "semanticDiag");
      if (!found) {
        throw new assert.AssertionError();
      }
      expect(maskFileNameForDiagnostics(found)).toMatchSnapshot();
    });
  });

  describe("#getCodeFixes", () => {
    it("should return codeFixes reported by ESLint rules", async () => {
      server = createServer({ projectPath: path.resolve(__dirname, "../projects/simple") });
      const { file, fileContent } = server.readFile("./main.ts");
      server.send({ command: "open", arguments: { file, fileContent, scriptKindName: "TS" } });
      await server.waitEvent("projectLoadingFinish");
      server.send({ command:"getCodeFixes",arguments:{ file, startLine:1, startOffset:12, endLine:1, endOffset:12, errorCodes:[30010] } });
      await server.waitResponse("getCodeFixes");
      const found = findCommandResponse(server.responses, "getCodeFixes");
      if (!found) {
        throw new assert.AssertionError();
      }
      expect(maskFileNameForCodeFixes(found)).toMatchSnapshot();
    });
  });

});
