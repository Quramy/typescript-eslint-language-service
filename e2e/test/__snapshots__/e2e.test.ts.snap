// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageService plugin #getCodeFixes should return codeFixes reported by ESLint rules 1`] = `
Object {
  "body": Array [
    Object {
      "changes": Array [
        Object {
          "fileName": "<fileName>",
          "textChanges": Array [
            Object {
              "end": Object {
                "line": 1,
                "offset": 12,
              },
              "newText": ";",
              "start": Object {
                "line": 1,
                "offset": 12,
              },
            },
          ],
        },
      ],
      "description": "Fix: Missing semicolon.",
      "fixId": "semi",
      "fixName": "<fileName>",
    },
  ],
  "command": "getCodeFixes",
  "request_seq": 2,
  "seq": 0,
  "success": true,
  "type": "response",
}
`;

exports[`LanguageService plugin #getSemanticDiagnostics should not reproduce issue #7 1`] = `
Object {
  "body": Object {
    "diagnostics": Array [],
    "file": "<file>",
  },
  "event": "semanticDiag",
  "seq": 0,
  "type": "event",
}
`;

exports[`LanguageService plugin #getSemanticDiagnostics should not return ESLint error when the project does not use @typescript-eslint/parser 1`] = `
Object {
  "body": Object {
    "diagnostics": Array [],
    "file": "<file>",
  },
  "event": "semanticDiag",
  "seq": 0,
  "type": "event",
}
`;

exports[`LanguageService plugin #getSemanticDiagnostics should return ESLint error when the project is configured with ESLint plugins 1`] = `
Object {
  "body": Object {
    "diagnostics": Array [
      Object {
        "category": "error",
        "code": 30010,
        "end": Object {
          "line": 2,
          "offset": 10,
        },
        "start": Object {
          "line": 2,
          "offset": 9,
        },
        "text": "[@typescript-eslint/no-unused-vars] 'a' is assigned a value but never used.",
      },
    ],
    "file": "<file>",
  },
  "event": "semanticDiag",
  "seq": 0,
  "type": "event",
}
`;

exports[`LanguageService plugin #getSemanticDiagnostics should return ESLint error when the project uses @typescript-eslint/parser 1`] = `
Object {
  "body": Object {
    "diagnostics": Array [
      Object {
        "category": "error",
        "code": 30010,
        "end": Object {
          "line": 1,
          "offset": 13,
        },
        "start": Object {
          "line": 1,
          "offset": 12,
        },
        "text": "[semi] Missing semicolon.",
      },
    ],
    "file": "<file>",
  },
  "event": "semanticDiag",
  "seq": 0,
  "type": "event",
}
`;
